File: src/cli/ainarictl/resources/instance.go
/**
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022-2026 Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

package ainari_resources

import (
	"fmt"
	ainarictl_common "ainarictl/common"
	"os"

	ainari_sdk "github.com/kitsudaiki/ainari"
	"github.com/spf13/cobra"
)

var (
	templatePath   string
	checkpointName string
	instanceMode    string
)

var createInstanceCmd = &cobra.Command{
	Use:   "create -t TEMPLATE_PATH NAME",
	Short: "Create a new instance.",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		context, err := Login()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		instanceName := args[0]
		templateContent, err := os.ReadFile(templatePath)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		content, err := ainari_sdk.CreateInstance(context, instanceName, string(templateContent))
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		ainarictl_common.PrintSingle(content)
	},
}

var getInstanceCmd = &cobra.Command{
	Use:   "get CLUSTER_UUID",
	Short: "Get information of a specific instance.",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		context, err := Login()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		instanceUuid := args[0]
		content, err := ainari_sdk.GetInstance(context, instanceUuid)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		ainarictl_common.PrintSingle(content)
	},
}

var listInstanceCmd = &cobra.Command{
	Use:   "list",
	Short: "List all instance.",
	Run: func(cmd *cobra.Command, args []string) {
		context, err := Login()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		content, err := ainari_sdk.ListInstance(context)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		ainarictl_common.PrintList(content["instances"].([]interface{}))
	},
}

var deleteInstanceCmd = &cobra.Command{
	Use:   "delete CLUSTER_UUID",
	Short: "Delete a specific instance from the backend.",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		context, err := Login()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		instanceUuid := args[0]
		_, err = ainari_sdk.DeleteInstance(context, instanceUuid)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		fmt.Printf("successfully deleted instance '%v'\n", instanceUuid)
	},
}

var instanceCmd = &cobra.Command{
	Use:   "instance",
	Short: "Manage instance.",
}

func Init_Instance_Commands(rootCmd *cobra.Command) {
	rootCmd.AddCommand(instanceCmd)

	instanceCmd.AddCommand(createInstanceCmd)
	createInstanceCmd.Flags().StringVarP(&templatePath, "template", "t", "", "Instance Template (mandatory)")
	createInstanceCmd.MarkFlagRequired("template")

	instanceCmd.AddCommand(getInstanceCmd)

	instanceCmd.AddCommand(listInstanceCmd)

	instanceCmd.AddCommand(deleteInstanceCmd)
}

File: src/libs/rust/ainari_clients/src/instance.rs
// Copyright 2022-2026 Tobias Anker <tobias.anker@kitsunemimi.moe>

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use uuid::Uuid;

use ainari_api_structs::instance_structs::*;
use ainari_common::error::AinariError;
use ainari_common::secret::Secret;

use crate::prepare_client;
use crate::{handle_empty_response, handle_response};

/// Creates a new instance in the Ainari system.
///
/// # Arguments
///
/// * `sakura_address` - The base URL of the Ainari Sakura service.
/// * `token` - Authentication token for the API.
/// * `internal_api_key` - Internal API key for authorization.
/// * `name` - Name of the instance to be created.
/// * `template` - Template to be used for instance creation.
/// * `insecure_client` - Whether to use an insecure client (no TLS verification).
///
/// # Returns
///
/// A `Result` containing the created `InstanceResp` on success, or an `AinariError` on failure.
pub async fn create_instance(
    sakura_address: &String,
    token: &String,
    internal_api_key: &Secret,
    name: &str,
    template: &str,
    insecure_client: bool,
) -> Result<InstanceResp, AinariError> {
    let client = prepare_client(sakura_address, insecure_client);
    let url = format!("{sakura_address}/v1alpha/instance/internal");

    // Create the request body with the provided name and template
    let body = InstanceCreateReq {
        template: template.to_owned(),
        name: name.to_owned(),
    };
    let json_str = serde_json::to_string(&body).unwrap();

    // Send the POST request to create the instance
    let response = client
        .post(url)
        .insert_header(("Authorization", format!("Bearer {}", token)))
        .insert_header(("X-Internal-API-Key", internal_api_key.reveal()))
        .insert_header(("Content-Type", "application/json"))
        .send_body(json_str)
        .await;

    // Handle the response and return the result
    let resp: Result<InstanceResp, AinariError> = handle_response(response, "instance", "").await;
    resp
}

/// Retrieves information about a specific instance from the Ainari system.
///
/// # Arguments
///
/// * `sakura_address` - The base URL of the Ainari Sakura service.
/// * `token` - Authentication token for the API.
/// * `internal_api_key` - Internal API key for authorization.
/// * `instance_uuid` - UUID of the instance to retrieve.
/// * `insecure_client` - Whether to use an insecure client (no TLS verification).
///
/// # Returns
///
/// A `Result` containing the retrieved `InstanceResp` on success, or an `AinariError` on failure.
pub async fn get_instance(
    sakura_address: &String,
    token: &String,
    internal_api_key: &Secret,
    instance_uuid: &Uuid,
    insecure_client: bool,
) -> Result<InstanceResp, AinariError> {
    let client = prepare_client(sakura_address, insecure_client);
    let url = format!("{sakura_address}/v1alpha/instance/{instance_uuid}/internal");

    // Send the GET request to retrieve the instance information
    let response = client
        .get(url)
        .insert_header(("Authorization", format!("Bearer {}", token)))
        .insert_header(("X-Internal-API-Key", internal_api_key.reveal()))
        .send()
        .await;

    // Handle the response and return the result
    let resp: Result<InstanceResp, AinariError> =
        handle_response(response, "instance", &instance_uuid.to_string()).await;
    resp
}

/// Lists all instances available in the Ainari system.
///
/// # Arguments
///
/// * `sakura_address` - The base URL of the Ainari Sakura service.
/// * `token` - Authentication token for the API.
/// * `internal_api_key` - Internal API key for authorization.
/// * `insecure_client` - Whether to use an insecure client (no TLS verification).
///
/// # Returns
///
/// A `Result` containing the list of instances as `InstanceListResp` on success, or an `AinariError` on failure.
pub async fn list_instance(
    sakura_address: &String,
    token: &String,
    internal_api_key: &Secret,
    insecure_client: bool,
) -> Result<InstanceListResp, AinariError> {
    let client = prepare_client(sakura_address, insecure_client);
    let url = format!("{sakura_address}/v1alpha/instance/internal");

    // Send the GET request to list all instances
    let response = client
        .get(url)
        .insert_header(("Authorization", format!("Bearer {}", token)))
        .insert_header(("X-Internal-API-Key", internal_api_key.reveal()))
        .send()
        .await;

    // Handle the response and return the result
    let resp: Result<InstanceListResp, AinariError> = handle_response(response, "instance", "").await;
    resp
}

/// Deletes a specific instance from the Ainari system.
///
/// # Arguments
///
/// * `sakura_address` - The base URL of the Ainari Sakura service.
/// * `token` - Authentication token for the API.
/// * `internal_api_key` - Internal API key for authorization.
/// * `instance_uuid` - UUID of the instance to delete.
/// * `insecure_client` - Whether to use an insecure client (no TLS verification).
///
/// # Returns
///
/// A `Result` indicating success or failure. On success, returns `Ok(())`.
pub async fn delete_instance(
    sakura_address: &String,
    token: &String,
    internal_api_key: &Secret,
    instance_uuid: &Uuid,
    insecure_client: bool,
) -> Result<(), AinariError> {
    let client = prepare_client(sakura_address, insecure_client);
    let url = format!("{sakura_address}/v1alpha/instance/{instance_uuid}/internal");

    // Send the DELETE request to remove the instance
    let response = client
        .delete(url)
        .insert_header(("Authorization", format!("Bearer {}", token)))
        .insert_header(("X-Internal-API-Key", internal_api_key.reveal()))
        .send()
        .await;

    // Handle the empty response and return the result
    handle_empty_response(response, "instance", &instance_uuid.to_string()).await
}
